torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -640,6 +681,172 @@ def _need_prefetch_post_backward_hook(self) -> bool:
         else:
             return False
 
+    @contextlib.contextmanager
+    def set_state_dict_type(self, state_dict_type: StateDictType) -> Generator:
+        """
+        A context manager to set the state_dict_type of this FSDP module and
+        its descendant FSDP modules.
+        .. note:: This API should be called for only the root FSDP module.
+        .. note:: All the FSDP modules should have the same state_dict_type.
+        .. note:: The default state_dict_type is StateDictTyp.STATE_DICT.
+
+        Args:
+            state_dict_type (StateDictType): the desired state_dict_type to set.
+        """
+        prev_state_dict_type = self.state_dict_type
+        for module in self.modules():
+            if isinstance(module, FullyShardedDataParallel):
+                if module.state_dict_type != prev_state_dict_type:
+                    raise RuntimeError(
+                        "All FSDP module should the same state_dict_type."
+                    )
+                module.state_dict_type = state_dict_type
+        try:
+            yield
+        finally:
+            for module in self.modules():
+                if isinstance(module, FullyShardedDataParallel):
+                    module.state_dict_type = prev_state_dict_type
+
+    @staticmethod
+    def _post_state_dict_hook(
```

Comment:
Should we use some sort of basic dispatch mechanism to deconvolute / separate out implementation details from this code? 

Maybe something like 3 functions:

`def _local_post_state_dict_hook`
`def _full_post_state_dict_hook`
`def _sharded_state_dict_hook`

And in `_post_state_dict_hook`, we have:
```
dispatch: Dict[StateDictType, Callable] = { 
    StateDictType.LOCAL: _local_post_state_dict_hook,
    # and so on
}
```

and the implementation is just `return dispatch[state_dict_type](args)`

eventually, those 3 functions might be able to be moved into some sort of utils folder.